optimum/neuron/modeling.py

```
@@ -506,3 +511,136 @@ def forward(
+    Neuron model with a causal language modeling head for inference on Neuron devices.
+    """
+
+    auto_model_class = AutoModelForCausalLM
+    main_input_name = "input_ids"
+
+    def __init__(self, model, config, model_path, generation_config):
+        super().__init__(model, config, model_path, generation_config)
+        self.cur_len = 0
+        neuron_config = getattr(config, "neuron", None)
+        if neuron_config is None:
+            raise ValueError(
+                "The specified directory does not contain a neuron model. "
+                "Please convert your model to neuron format by passing export=True."
+            )
+        self.batch_size = neuron_config["neuron_kwargs"]["batch_size"]
+        self.max_length = neuron_config["neuron_kwargs"]["n_positions"]
+
+    def reset_generation(self):
+        self.cur_len = 0
+
+    @add_start_docstrings_to_model_forward(
+        NEURON_CAUSALLM_MODEL_DOCSTRING
+        + TEXT_GENERATION_EXAMPLE.format(
+            processor_class="AutoTokenizer",
+            model_class="NeuronModelForCausalLM",
+            checkpoint="gpt2",
+        )
+    )
+    def forward(
+        self,
+        input_ids,
+        cache_ids,
+        start_ids=None,
+        output_hidden_states=False,
+        output_attentions=False,
+        attention_mask=None,
+        return_dict=False,
+    ):
+        if output_hidden_states or output_attentions or attention_mask is not None:
+            warnings.warn(
+                "Warning: These arguments are not used by forward(): \
+                (output_hidden_states, output_attentions, attention_mask)"
+            )
+        # Evaluate the output logits, storing the current key and values at the indices specified by cache_ids
+        out_logits = self.model.forward(input_ids, cache_ids, start_ids)
+        out_logits = out_logits[:, None, :]
+        # Since we are using a static cache, we don't need to return past keys and values
+        if return_dict:
```

AFAIK, in optimum, we always `return_dict`, maybe just keep the convention.

---

OK, so do I need to raise an error or add a warning and ignore if return_dict=False ?

---

@dacorvo Personally don't feel it necessary, what about just remove `return_dict` from input arguments?

---

I tried to remove it, but I got an error and remembered why I had to put it there in the first place: this is because generate will pass it no matter what.

---

Alright, I see, for onnxruntime it seems that someone added `**kwargs` to let it pass which I feel not the optimal solution either... but why not, let's either adopt that solution or set the default value to `True` at least?

---

Done